### PR TITLE
Bump targetSdkVersion to API 31

### LINKED ISSUE #5129: Bump targetSdkVersion to API 31
#### Problem description
Every year we have to bump the `targetSdkVersion` to [Meet Google Play's target API level requirement](https://developer.android.com/google/play/requirements/target-sdk).
Theoretically we can keep releasing new versions of our app till November with the current `targetSdkVersion` which is 30 but I think this year we should do that earlier becasue:
- I remember that last year we procrastinated and the conclusion was that next time we should do that earlier
- It's better to do that at the begging of a release cycle to have time to catch bugs that might occur 
- It might be better to do that now when @kkrawczyk123 is still available and can help @srujner with this crazy annual case where anything might be broken

#### Expected behavior
The app should work well with `targetSdkVersion` 31 with no bugs/crashes/regressions

#### Things to do:
- [x] Bump `targetSdkVersion` to 31
- [x] Explicitly declare the [android:exported](https://developer.android.com/guide/topics/manifest/activity-element#exported) attribute for components that should be exported https://developer.android.com/about/versions/12/behavior-changes-12#exported
- [x] Specify pending intents mutability https://developer.android.com/about/versions/12/behavior-changes-12#pending-intent-mutability
- [x] Double check toast text will fit https://developer.android.com/about/versions/12/behavior-changes-12#toast-redesign
- [x] Check behaviour of permissions and scheduled jobs (auto update, match exactly etc) when coming in and out of hibernation https://developer.android.com/topic/performance/app-hibernation#manually-invoke

@seadowg please when you are free also review the documentation https://developer.android.com/about/versions/12/behavior-changes-12 to make sure we do not miss anything obvious.

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #5129 

#### What has been done to verify that this works as intended?
I've reviewed the documentation https://developer.android.com/about/versions/12/behavior-changes-12#toast-redesign and implemented changes we identified in #5129 
I've also run tests on Android 12.

#### Why is this the best possible solution? Were any other approaches considered?
This is what we have to do every year as I discussed i the issue.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
This might affect the app in any way. We have tried to identify as many features that might be broken as possible and fix it but this kind of change (bumping `targetSdkVersion` is tricky and it's not possible to list everything that should be tested).
Some ideas are:
- Toasts - they will contain the icon on Android 12 and they length will be restricted to two lines (see https://developer.android.com/about/versions/12/behavior-changes-12#toast-redesign)
- Testing the app on large screens and multi window mode (see https://developer.android.com/about/versions/12/behavior-changes-all#multi-window-mode)
- Camera preview on large screens (see https://developer.android.com/about/versions/12/behavior-changes-all#camera-preview)
- Microphone and camera indicators (see https://developer.android.com/about/versions/12/behavior-changes-all#multi-window-mode)

Generally it would be good for the QA team to read the whole doc:
- https://developer.android.com/about/versions/12/behavior-changes-all
- https://developer.android.com/about/versions/12/behavior-changes-12

to help you understand what kid of changes might occur while testing app on newer android versions (in this case android 12) and after bumping `targetSdkVersion`.

@seadowg if you have any other ideas please share.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

**Review by seadowg — APPROVED** [review]

**Comment by srujner** [comment]
@grzesiek2010 ODK collect is crashing after clicking on the "Start GeoPoint" button for geopoint type with no appearance in All widgets. It is reproducible for all map sources only on Android 12. 
Here are some logs: 
crash1
crash2

**Comment by grzesiek2010** [comment]
> ODK collect is crashing after clicking on the "Start GeoPoint" button for geopoint type

It should be fixed now. The problem was in observing the number of available satellites which we display in the dialog:
Screenshot_1652427379

I had to use a different provider to achieve that. Unfortunately I wasn't able to use it for all android versions because the new one is available on Android 7+ only. So for android 5-6 the old provider should be used and for android 7+ the new one which I've just implemented. You shouldn't see any differences during testing it's something that works under the hood but I'm just bringing it up to let you know that theoretically not only android 12 but all versions 7+ might be affected and should be tested to confirm that the number of satellites is displayed correctly.

PS this is the nature of bumping sdk versions. We knew the old provider was deprecated and we should replace it sooner or later but it wasn't mentioned anywhere that it will cause crashes on Android 12.

File: location/src/main/java/org/odk/collect/location/satellites/GpsStatusSatelliteInfoClient.kt [hunk 1]
```diff
@@ -35,3 +43,30 @@ private class GpsStatusSatellitesLiveData(private val locationManager: LocationM
...
 }
+
+@SuppressLint("MissingPermission")
+@RequiresApi(Build.VERSION_CODES.N)
+private class GnssStatusSatellitesLiveData(private val locationManager: LocationManager) :
+    NonNullLiveData<Int>(0) {
+
+    private val gnssStatusCallback = object : GnssStatus.Callback() {
+        override fun onSatelliteStatusChanged(status: GnssStatus) {
+            super.onSatelliteStatusChanged(status)
```
Comment (seadowg) [inline location/src/main/java/org/odk/collect/location/satellites/GpsStatusSatelliteInfoClient.kt]: I'm guessing you don't need this super call?

File: location/src/main/java/org/odk/collect/location/satellites/GpsStatusSatelliteInfoClient.kt [hunk 2]
```diff
@@ -35,3 +43,30 @@ private class GpsStatusSatellitesLiveData(private val locationManager: LocationM
...
+            var usedInFix = 0
```
Comment (seadowg) [inline location/src/main/java/org/odk/collect/location/satellites/GpsStatusSatelliteInfoClient.kt]: Not essential (and mostly subjective), but you can avoid needing any mutation here with a `fold`:

```kotlin
value = (0 until status.satelliteCount).fold(0) { count, index ->
    when {
        status.usedInFix(index) -> count + 1
       else -> count
    }
}
```

File: location/src/main/java/org/odk/collect/location/satellites/GpsStatusSatelliteInfoClient.kt (hunk 1 above)
Comment (grzesiek2010) [reply location/src/main/java/org/odk/collect/location/satellites/GpsStatusSatelliteInfoClient.kt]: Right

File: location/src/main/java/org/odk/collect/location/satellites/GpsStatusSatelliteInfoClient.kt (hunk 2 above)
Comment (grzesiek2010) [reply location/src/main/java/org/odk/collect/location/satellites/GpsStatusSatelliteInfoClient.kt]: Done.

**Review by seadowg — APPROVED** [review]

**Comment by srujner** [comment]
**Tested with Success!**
Verified on Android 10 and 12

Verified cases:

1. Maps:
- Regression checks for GeoPoint, GeoTrace, GeoShape widgets;
- Select Map forms
- Offline Layers;
- Google Maps, Mapbox, OSM;
- Collecting Location in background;
- View Submissions on map;
- Screen rotation, minimize app, using backward button,

2. Forms
- Sending finalized Forms, Editing saved Forms, Viewing sent Forms, Deleting saved Forms;
- All notifications: Tapping on notification, clicking on show details, exit from the error list;
- Match exactly and Failed match exactly;
- Form version and media updates;
- Automatic download and Failed automatic download,

3. Settings
- Regression checks and Exploratory testing on Project settings and user settings;
- Adding project manually or through QR code;
- Switching through projects, adding duplicated projects;
- Update settings from older version;
- Invalid/Inverted QR code,

4. Camera
- Regression checks on Audio, Video and Image widgets;
- Using external app for audio recording, background recording;
- Using Front camera, selfie widget;
- Adding big pictures and using different image size;
- Checking all file formats
- Screen rotation, minimize app, using backward button,

5. Toast
- Displaying Toast messages;
- Displaying ODK Collect icon in toast messages;
- Displaying Toast messages longer than 2 lines;
- Toasts checked: save form, offline mode, form without instanceID, switch between projects, missing external app toast, invalid URL on fallback submission URL and server URL, invalid email address, experimental settings, set admin password, edit admin password, invalid admin password, disable admin password, scan invalid qr code, 

6. Other things
- Regression checks on All Widgets.xml;
- Exploratory testing;
- Permission dialogs;
- Project Management and Access control settings;
- Light and Dark theme.

**Comment by kkrawczyk123** [comment]
Verified also on Androids: 5.1 and 8.1.

**Pull request merged**